Optionally allow redefinition of variable with different type

JukkaL opened the pull request:
If the flag `--allow-redefinition` is used, it's now possible to redefine a 
variable with a different type. For example, this code will be okay:

```
x = 0  # Define 'x' with type 'int'
print(x)
x = 'a'  # Define another 'x' with type 'str'
print(x)
```

The variable needs to be read before redefinition is possible. This is mainly 
to allow these idioms:

```
x: List[int]
x = []  # Not a redefinition

y = None  # type: List[int]  # Strict optional disabled
y = []  # Not a redefinition
```

It's also okay to use a type annotation in any of the assignments. This is
fine:

```
x: int = 0
print(x)
x: str = 'a'
print(x)
```

Redefinition is only allowed to happen in the same block and the nesting level
as the original definition. This lets us keep the implementation relatively simple,
and there will hopefully be less of a readability impact, since it's less likely that
users accidentally redefine a variable or use redefinition in confusing ways. We 
may want to lift the restriction in the future, however. Function arguments can
be redefined in the body but only in non-nested blocks.

The implementation does a simple static analysis to rename variables to make
them distinct when we detect redefinition. This happens before semantic 
analysis pass 1.

The new behavior is not fully backward compatible with the old behavior. For
example, here we lose type context in the second assignment:

```
x: List[int] = []
f(x)
x = []  # Need type annotation since this creates an independent variable
```

Internally we use a sequence of single quotes as a variable suffix
to represent renamed variants. The implementation provides an `unmangle()`
function to get the original name from a potentially mangled name.

It's still impossible to redefine final names, classes and certain other things.
The main use case is simple variables, either in single or multiple assignments.

I did some benchmarking (not with the latest version though) and the performance 
impact of the new pass was minor (under 2% IIRC).

Things to do in additional PRs:
* Add documentation.
* Make sure we unmangle names everywhere.
* Consider allowing redefinition to happen sometimes in another block.

ilevkivskyi commented on test-data/unit/check-redefine.test:
```diff
@@ -0,0 +1,451 @@
+    _ = 0
+    _ = ''
+    x, _ = 0, C()
+[builtins fixtures/tuple.pyi]
```
I would add test cases for
```python
e = 1
e
try:
    ...
except Exception as e:
    ...
```
and
```python
with open('file') as f:  # IO[str]
    ...
with open('file', 'b') as f:  # IO[bytes]
    ...
```
(maybe plus some variations).

JukkaL replied:
Added test cases.

JukkaL commented on mypy/renaming.py:
```diff
@@ -0,0 +1,347 @@
+        super().visit_try_stmt(stmt)
+        self.leave_with_or_try()
+
+    def visit_with_stmt(self, stmt: WithStmt) -> None:
```
Try/except has special handling in the type checker, so it's not needed (added comment). Fixed with statements.

JukkaL commented on mypy/renaming.py:
```diff
@@ -0,0 +1,347 @@
+
+    def handle_arg(self, name: str) -> None:
+        """Store function argument."""
+        if name not in self.refs[-1]:
```
Removed the check.

JukkaL commented on mypy/renaming.py:
```diff
@@ -0,0 +1,347 @@
+
+        This will be called at the end of a scope.
+        """
+        is_func = self.is_nested()
```
Fixed this.

JukkaL commented on test-data/unit/check-final.test:
```diff
@@ -336,14 +336,15 @@ class C:
 def f() -> int:
     global x
-    x = 3  # E: Cannot assign to final name "x"
+    x = 3  # No error here is okay since we reported an error above
```
Yes. Added a test case.

JukkaL commented:
I think that I've now addressed all the feedback.

Agreed that this will need to be documented by the next release (or we should hide the option).

> Maybe when we detect UninhabitedType, instead of immediately giving an error we can try to re-infer type of r.h.s. using type of the "previous" x (with one prime less) as a context?

I thought about this but I couldn't come up with a simple implementation, so I decided to postpone this. Calculating the previous 'x' is more complicated, since the rules for renaming are different for globals and locals, for example. This is a potential future improvement. I'll create a follow-up issue about this.

neighthan commented:
Is there a way to only allow redefinition if an explicit type is given? So that 

```python
x = 0
print(x)
x: str = 'a'
print(x)
```

is valid but

```python
x = 0
print(x)
x = 'a'
print(x)
```

is not? This would prevent any accidental redefinition and make it clear that the type is supposed to be changing (in the example above it's clear anyway, but if we had `x = some_func()` it might not be obvious anymore).

oakkitten commented:
this doesn't seem to work for `def`:

```
from typing import Callable, cast

foo: str = "a"
foo: int = cast(int, foo)

bar: Callable[[], int] = lambda: 1
bar: int = cast(int, bar)

def baz() -> int:
    return 1
baz: int = cast(int, baz)
```
> 11: error: Name 'baz' already defined on line 9

any workarounds?
